shared/src/ledger/eth_bridge/storage/bridge_pool.rs

```diff
@@ -44,3 +58,837 @@ pub fn is_bridge_pool_key(key: &Key) -> bool {
+                        next_hashes
+                            .push(Node::OnPath(hash_pair(hash.clone(), sib)));
+                    }
+                    (Node::Sibling(sib), Node::OnPath(hash)) => {
+                        flags.push(false);
+                        proof_hashes.push(sib.clone());
+                        next_hashes
+                            .push(Node::OnPath(hash_pair(hash, sib.clone())));
+                    }
+                    (Node::Sibling(left), Node::Sibling(right)) => {
+                        next_hashes.push(Node::Sibling(hash_pair(
+                            left.clone(),
+                            right,
+                        )));
+                    }
+                }
+            }
+            hashes = next_hashes;
+        }
+        // add the root to the proof
+        if flags.is_empty() && proof_hashes.is_empty() && leaves.is_empty() {
+            proof_hashes.push(self.root.clone());
+        }
+
+        Ok(BridgePoolProof {
+            proof: proof_hashes,
+            leaves: values,
+            flags,
+        })
+    }
+
+    /// Parse a db key to see if it is valid for the
+    /// bridge pool.
+    ///
+    /// It should have one string segment which should
+    /// parse into a [Hash]
+    fn parse_key(key: &Key) -> Result<KeccakHash, Error> {
+        if key.segments.len() == 1 {
+            match &key.segments[0] {
+                DbKeySeg::StringSeg(str) => {
+                    str.as_str().try_into().map_err(|_| {
+                        eyre!("Could not parse key segment as a hash").into()
+                    })
+                }
+                _ => Err(eyre!("Bridge pool keys should be strings.").into()),
+            }
+        } else {
+            Err(eyre!(
+                "Key for the bridge pool should have exactly one segment."
+            )
+            .into())
+        }
+    }
+}
+
+/// Concatenate two keccak hashes and hash the result
+#[inline]
+fn hash_pair(left: KeccakHash, right: KeccakHash) -> KeccakHash {
+    if left.0 < right.0 {
+        keccak_hash([left.0, right.0].concat().as_slice())
+    } else {
+        keccak_hash([right.0, left.0].concat().as_slice())
+    }
+}
+
+/// Keeps track if a node is on a path from the
+/// root of the merkle tree to one of the leaves
+/// being included in a multi-proof.
+#[derive(Debug, Clone)]
+enum Node {
+    /// Node is on a path from root to leaf in proof
+    OnPath(KeccakHash),
+    /// Node is not on a path from root to leaf in proof
+    Sibling(KeccakHash),
+}
+
+impl Default for Node {
+    fn default() -> Self {
+        Self::Sibling(Default::default())
+    }
+}
+
+/// A multi-leaf membership proof
+pub struct BridgePoolProof {
+    /// The hashes other than the provided leaves
+    pub proof: Vec<KeccakHash>,
+    /// The leaves; must be sorted
+    pub leaves: Vec<PendingTransfer>,
+    /// flags to indicate how to combine hashes
+    pub flags: Vec<bool>,
```

Comment:
Could we document what the elements of `flags` are meant to mean, if they are `true` or `false`? Could even use an enum something like

```rust
enum Flag {
	BothNodesOnPath,
	OneNodeOnPath,
}
```

I'm guessing from looking at how we set `flags` in `get_membership_proof`

Reply:
These have to end up being bool for solidity, and I wanted to make sure that the test code in rust was as close a translation to the solidity code as possible. But I can add more doc strings about it.